src/mscorlib/src/System/Buffer.cs

```diff
@@ -401,87 +405,173 @@ internal unsafe static void Memmove(byte* dest, byte* src, uint len)
                 {
-                    *dest = *src;
-                    src++;
-                    dest++;
-                    len--;
-                    if (((int)dest & 2) == 0)
-                        goto Aligned;
+                    *(dest + i) = *(src + i);
+                    i += 1;
+                    if (((int)dest & 2) != 0)
+                        goto IntAligned;
                 }
-                *(short *)dest = *(short *)src;
-                src += 2;
-                dest += 2;
-                len -= 2;
-            Aligned: ;
+                *(short*)(dest + i) = *(short*)(src + i);
+                i += 2;
             }
 
+            IntAligned:
+
 #if BIT64
-            if (((int)dest & 4) != 0)
+            if ((((int)dest - 1) & 4) == 0)
             {
-                *(int *)dest = *(int *)src;
-                src += 4;
-                dest += 4;
-                len -= 4;
+                *(int*)(dest + i) = *(int*)(src + i);
+                i += 4;
             }
 #endif
 
-#if BIT64
-            ulong count = len / 16;
-#else
-            uint count = len / 16;
-#endif
-            while (count > 0)
+            nuint end = len - 16;
+
+            // We know due to the above switch-case that
+            // this loop will always run 1 iteration; max
+            // bytes we copy before checking is 23 (7 to
+            // align the pointers, 16 for 1 iteration) so
+            // the switch handles lengths 0-22.
+            Contract.Assert(end >= 7 && i <= end);
+
+            do
             {
+                // This loop looks very costly since there
+                // appear to be a bunch of temporary values
+                // being created with the adds, but the jit
+                // (for x86 anyways) will convert each of
+                // these to use memory addressing operands.
+                // So the only cost is a bit of code size,
+                // which is made up for by the fact that
+                // we save on writes to dest/src.
+
 #if BIT64
-                ((long*)dest)[0] = ((long*)src)[0];
-                ((long*)dest)[1] = ((long*)src)[1];
+                *(long*)(dest + i) = *(long*)(src + i);
+                *(long*)(dest + i + 8) = *(long*)(src + i + 8);
 #else
-                ((int*)dest)[0] = ((int*)src)[0];
-                ((int*)dest)[1] = ((int*)src)[1];
-                ((int*)dest)[2] = ((int*)src)[2];
-                ((int*)dest)[3] = ((int*)src)[3];
+                *(int*)(dest + i) = *(int*)(src + i);
+                *(int*)(dest + i + 4) = *(int*)(src + i + 4);
+                *(int*)(dest + i + 8) = *(int*)(src + i + 8);
+                *(int*)(dest + i + 12) = *(int*)(src + i + 12);
 #endif
-                dest += 16;
-                src += 16;
-                count--;
+
+                i += 16;
             }
+            while (i <= end);
+
+            len -= i; // len now represents how many bytes are left after the unrolled loop
```

Comment:
@benaadams `end` is just `len - 16`, so that would always be 16. I suppose this could come before the loop since i is always incremented by 16 however, so the lower 4 bits would still be the same.